lib/chrono_model/time_machine/history_model.rb

```diff
@@ -210,11 +210,11 @@ def record # :nodoc:
       end
 
       def valid_from
-        validity.first
+        validity.begin if validity.begin.is_a?(Time)
       end
 
       def valid_to
-        validity.last
+        validity.end if validity.end.is_a?(Time)
```

Comment:
This will nil for any other validity that is present but not a type?
Isnt this misleading if i access the range rather than the single valid_to - valid_from?

Reply:
Why do we need this conditional: `if validity.end.is_a?(Time)` @gridanjbf

Reply:
Sorry, maybe I did not get the question.

`validity` is a column with type `tsrange`.

In Rails, it is returned as a `Range` object (if present), and it may have `begin` and `end`, or one of the two if it is an endless/beginless range

```rb
3.2.3 :001 > (Time.now..).end.is_a?(Time)
 => false 
3.2.3 :002 > (..Time.now).end.is_a?(Time)
 => true 
3.2.3 :003 > (..Time.now).begin.is_a?(Time)
 => false 
```

When used in chronomodel, there is no practical difference between the old and the new approach:

```rb
# Custom tsrange parser
[3] pry(main)> City::History.new.validity
=> nil

[4] pry(main)> City::History.new.valid_from
# NoMethodError: undefined method `first' for nil

[5] pry(main)> City::History.new.valid_to
# NoMethodError: undefined method `last' for nil

# Default tsrange parser
[1] pry(main)> City::History.new.validity
=> nil

[2] pry(main)> City::History.new.valid_from
# NoMethodError: undefined method `begin' for nil

[3] pry(main)> City::History.new.valid_to
# NoMethodError: undefined method `end' for nil
```

With persisted chronomodel objects, there is no difference as well with `valid_from` and `valid_to`

```rb
# Custom parser
[2] pry(main)> City.last.history.last.valid_from
=> 2024-02-23 14:04:22.234841 UTC

[3] pry(main)> City.last.history.last.valid_to
=> nil

# Default parser
[5] pry(main)> City.last.history.last.valid_from
=> 2024-02-23 14:04:22.234841 UTC

[6] pry(main)> City.last.history.last.valid_to
=> nil
```

This _may_ be confusing if you expect Infinity, but:
- preserves the current behavior of Chronomodel (returning `nil` instead of Infinity)
- preserves default behavior of `tsrange` (returning a `Range` rather than an `Array`)
- allows Rails 7.0 to keep working (because 7.0 is not able to convert Infinity to timestamps)
- `valid_from` and `valid_to` are not Rails' standard, so we can return the object that is needed by chronomodel

Hope this can clarify

Reply:
> Why do we need this conditional: `if validity.end.is_a?(Time)` @gridanjbf

this was my question
@tagliala 


My concern was about this:

Date.new.is_a?(Time)
=> false
development: main:0> Date.new..Date.tomorrow
=> Mon, 01 Jan -4712..Sat, 24 Feb 2024

So we are going to you use Time to construct validaty range.

Reply:
Got it,  but as far as I can understand, `tsrange` has only `Time`stamps. Date would be a `daterange` field 

https://www.postgresql.org/docs/current/rangetypes.html#RANGETYPES-BUILTIN

I will clarify also this in the commit message

Reply:
Added comment and info to the commit message